### PR TITLE
Add validation for duplicate xml elements

### PR DESCRIPTION
**Pull request by oturpe**
Draft fix of #401 

As an example, duplicate checking for two elements is added. Using this simple system, checking all elements requires replicating the code for each checked field. Solution with less replication would require a new data structure and passing it around: instead of enumerated problems like `AS_APP_PROBLEM_DUPLICATE_PROJECT_LICENSE`, some kind of generic duplicate problem entry that takes element name as input.

Posting this draft for review, because I am really new to this codebase and cannot say what would be best way to proceed:
1. Replicate the simple solution for all fields that can only appear once
2. Develop the generic solution
3. Only check `project_license` duplicates, as that is what was reported in #401 and there the root of the problem is that even if a project may be under multiple licenses, there can only be one license field and separate syntax for encoding everything in there must be used.

Comments welcome.

### CODE REVIEW

File: libappstream-glib/as-app-validate.c [hunk 1]
```diff
@@ -1525,6 +1525,16 @@ as_app_validate (AsApp *app, guint32 flags, GError **error)
 					     AS_PROBLEM_KIND_TAG_INVALID,
 					     "<updatecontact> should be <update_contact>");
 		}
+		if ((problems & AS_APP_PROBLEM_DUPLICATE_PROJECT_LICENSE) > 0) {
+            ai_app_validate_add(helper,
+            	         AS_PROBLEM_KIND_TAG_INVALID,
+            	         "<project_license> was duplicated");
```
Comment (hughsie) [inline libappstream-glib/as-app-validate.c:1531]: intenting looks funky, 8 width tabs please

File: libappstream-glib/as-app.c [hunk 2]
```diff
@@ -5154,15 +5154,21 @@ as_app_node_parse_child (AsApp *app, GNode *n, guint32 flags,
 			priv->problems |= AS_APP_PROBLEM_TRANSLATED_LICENSE;
 			break;
 		}
+		if (priv->project_license != NULL) {
+     		priv->problems |= AS_APP_PROBLEM_DUPLICATE_PROJECT_LICENSE;
+	    }
```
Comment (hughsie) [inline libappstream-glib/as-app.c:5159]: indenting too

Comment (hughsie) [reply libappstream-glib/as-app.c:5159]: also no braces required

File: libappstream-glib/as-app.c [hunk 3]
```diff
@@ -5154,15 +5154,21 @@ as_app_node_parse_child (AsApp *app, GNode *n, guint32 flags,
...
-	/* <project_license> */
+	/* <metadata_license> */
 	case AS_TAG_METADATA_LICENSE:
 		if (as_node_get_attribute (n, "xml:lang") != NULL) {
 			priv->problems |= AS_APP_PROBLEM_TRANSLATED_LICENSE;
 			break;
 		}
+		if (priv->metadata_license != NULL) {
+     		priv->problems |= AS_APP_PROBLEM_DUPLICATE_METADATA_LICENSE;
+	    }
```
Comment (hughsie) [inline libappstream-glib/as-app.c:5171]: same

**Comment by hughsie** [comment]
A few nits, but looks basically what I would have done, so good job.

**Comment by oturpe** [comment]
Formatting fixed. A test case failed, apparently the "bad file" test already had a test file that had these problems, now the asserts need to be fixed. I will see to that soon.

**Comment by oturpe** [comment]
Tests added. In absence of further comments about which elements should be tested for duplicates, I left the scope as it was initially: only the license fields.

**Comment by oturpe** [comment]
Travis CI build failed, but to me that looks like an unrelated issue, as if something is generated without strict ordering, but is checked against fixed expectation:

```
Bail out! As:ERROR:../libappstream-builder/asb-self-test.c:524:asb_test_context_func: assertion failed (error == NULL):
--- /tmp/b	2021-05-17 19:42:30.984710347 +0000
+++ /tmp/a	2021-05-17 19:42:30.984710347 +0000
@@ -24,8 +24,8 @@
 <name>App</name>
 <summary>A test application</summary>
 <description><p>Long description goes here.</p></description>
-<icon type="cached" height="128" width="128">app.png</icon>
 <icon type="cached" height="64" width="64">app.png</icon>
+<icon type="cached" height="128" width="128">app.png</icon>
 <categories>
 <category>Profiling</category>
 <category>System</category>
 (-g-type-private--GTypeFlags, 0)
```

**Comment by hughsie** [comment]
> but to me that looks like an unrelated issue

Agreed! Could I be cheeky and ask you to add the fix for that to this branch please? It should just be reordering the expected lines in asb-self-test.c. Thanks!

**Comment by oturpe** [comment]
That error is now fixed, but another one was hidden behind it:

```
Bail out! As:ERROR:../libappstream-builder/asb-self-test.c:628:asb_test_context_func: assertion failed: (g_file_test ("/tmp/asbuilder/temp/icons/64x64/app.png", G_FILE_TEST_EXISTS))

--- stderr ---
**
As:ERROR:../libappstream-builder/asb-self-test.c:628:asb_test_context_func: assertion failed: (g_file_test ("/tmp/asbuilder/temp/icons/64x64/app.png", G_FILE_TEST_EXISTS))
-------
```

**Comment by hughsie** [comment]
> That error is now fixed, but another one was hidden behind it:

Okay, I can look at that later. Thanks for the fixes!

**Pull request merged**